src/billing_job.rs

```diff
@@ -0,0 +1,46 @@
+use std::io::Read;
+use std::sync::Arc;
+use std::time::Duration;
+
+use actix_web::web;
+use tiny_keccak::{Hasher, Keccak};
+use crate::model::AppState;
+use ethers::prelude::*;
+use ethers::core::abi::Abi;
+use k256::elliptic_curve::generic_array::sequence::Lengthen;
+
+pub async fn billing_scheduler(appstate: web::Data<AppState>,) -> Result<String, anyhow::Error> {
+    let abi_file_path = "src/contract_abi.json";          
+    let mut abi_json = String::new();
+    let mut file = std::fs::File::open(abi_file_path)?;
+    file.read_to_string(&mut abi_json)?;
+    let abi = serde_json::from_str::<Abi>(&abi_json)?;
+
+    let provider = Provider::<Http>::try_from(&appstate.rpc)?.interval(Duration::from_millis(1000));
+    let wallet: LocalWallet = appstate.operator_key.parse()?;
+
+    let client = SignerMiddleware::new(provider, wallet);
+    let contract = Contract::new(appstate.contract.as_str().parse::<Address>()?, abi, Arc::new(client));
+
+    let mut tx_hashes = Vec::new();
+    let mut amounts = Vec::new();
+    for (key, value) in appstate.service_costs.lock().unwrap().iter() {
```

Comment:
I was originally doing this only but that was causing an error while scheduling the job from main.rs related to "the trait `Send` is not implemented for `std::sync::MutexGuard<'_, HashMap<std::string::String, u128>>`
note: future is not `Send` as this value is used across an await". As far as I understood It basically is implying that the transaction for billing can't be sent if the mutex guard is still on.

Reply:
Only work around that I can think of is to empty out the costs and signature (also save these current values) and then drop the guard before sending the transaction. If somehow transaction couldn't be completed then we can update those mutex values with the previously saved costs and signature and throw the error out. 
Please share your views and let me know what is the best thing to do here.

Reply:
Use this https://docs.rs/tokio/latest/tokio/sync/struct.MutexGuard.html#method.mutex to resolve the error.

Reply:
> I was originally doing this only but that was causing an error while scheduling the job from main.rs related to "the trait `Send` is not implemented for `std::sync::MutexGuard<'_, HashMap<std::string::String, u128>>` note: future is not `Send` as this value is used across an await". As far as I understood It basically is implying that the transaction for billing can't be sent if the mutex guard is still on.

^ this error means something else entirely btw, it's basically saying `std::sync::MutexGuard` can't be sent across threads i.e. essentially not suitable for multi threaded env

Reply:
And while tokio mutex "solves" the issue, it's almost never a good idea to use it in place of std mutex - you will end up deadlocking the program or degrading the performance severely if you are not very careful. It's far better to try to solve the underlying issue instead.

Rust futures are basically giant state machines (try writing them by hand, it's a good exercise). Every `.await` is a suspend point for the state machine, and every time execution hits an `.await`, it stores info on where it is suspended. The next time the future is polled, it resumes from the previous suspend point. Now, what happens if you have variables that are declared before a suspend point but are also used after the point? The future has to store them so it can be restored when it resumes and used by code after the suspend point. Which is exactly what happens to your `MutexGuard` - "note: future is not Send as this value is used across an await". The future now stores the guard and cannot be used in a multithreaded env.

Note that `AppState.cgroups` just uses a std mutex but there are no errors. Difference is in how the mutex is never held across a suspend point and therefore is never part of the generated future.

Reply:
Ahh I think I got it now! Basically it's best to make sure that a mutex is never held across a suspend point, else the whole process will tend to execute synchronously causing slow performance (even with tokio::sync::Mutex). Thanks a lot for the explanation.